### ISSUE TITLE
ESQL: add nullable and multivalue information to the API result metadata

### ISSUE THREAD

**Issue by swallez** [elastic/elasticsearch#102697]
### Description

The `metadata` part of ESQL's JSON responses should provide information about whether a column's values can be null or multi-valued along with type information. AFAIU this information is readily available in the `Block` class.

Having this information is important to implement efficient parsing of responses, in particular in Elasticsearch Client Libraries. E.g. a non-nullable single-valued `long` column can be represented as an efficient `long[]`, something we can only do with the additional type information.

The format could be as follows:

```json5
{
  "columns": [
    {
      "name": "event.duration",
      "type": "long",
      "nullable": false,
      "multiple": false
    },
    ...
```

**Comment by nik9000** [comment]
We could tell you if any of the included values are null or an array, but
we couldn't tell you that, say, if you change a filter that they'll stay
that way. It's quite possible to change a date range and for fields to go
from non-nullable to nullable, for example.

Is that ok?

**Comment by swallez** [comment]
Yes, that would be fine, we don't need cross-request consistency.

There are two levels at which this information will be useful:
* at the application level, allowing users to introspect the types of a single response (similar to JDBC's `ResultSetMetadata`) if they implement some runtime type driven processing of the data
* at the client library level, allowing the library to choose the most optimal representation of a response's values

In both cases, this is about a single response.

**Comment by stratoula** [comment]
Thanx for pinging me Nik! 

> It's quite possible to change a date range and for fields to go
from non-nullable to nullable, for example.

@nik9000 I think we are fine. What we mostly care about is to distinguish fields that are nullable everywhere as they create a lot of noice in the apps.  I personally like the idea of having this information on the columns metadata. Especially the `nullable` could benefit Discover a lot. Also I think that the api proposed is simple so 👍 from me!

**Comment by costin** [comment]
This might work for now but it won't when adding pagination - a page might have nullable entries but the next one will not.
To rephrase, we know if a column is nullable or not ONLY when having the _whole_ results ; we do for now however that's something we want to avoid as it comes with limitations (such as the implicit limit).

**Comment by swallez** [comment]
Sorry for the late answer.

After much discussion in the Clients team, having nullability and multi-value information in the `columns` part even if it changes depending on filters and across pagination pages is still valuable for efficient parsing and in-memory representation of results as dataframes (i.e. columns).

The key here is to consider (and document) that the `columns` block defines the schema for the `values` block that follows, and only this one. In other words, a single ES|QL response body should be _internally consistent_, with no guarantees that its schema will be the same for all pages (at least for nullability and arity).

In the case of pagination, this is based on the assumption that the `columns` field will be output for every page, and not only at the beginning of a pagination sequence. I guess it's a safe assumption as it would otherwise require stateful clients that keep the schema from the first page even when doing deep pagination.
